File: jenjin-core/src/main/java/com/binarymonks/jj/physics/PhysicsRoot.java
package com.binarymonks.jj.physics;

import com.badlogic.gdx.math.Matrix3;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.Transform;
import com.binarymonks.jj.pools.N;

public interface PhysicsRoot {
    Vector2 position();

    float rotationR();

    /**
     * Be careful if you are using Box2D physics.
     * Rather apply forces through {@link PhysicsRoot#getB2DBody()}
     *
     * @param x
     * @param y
     */
    void setPosition(float x, float y);

    /**
     * Be careful if you are using Box2D physics.
     * Rather apply forces through {@link PhysicsRoot#getB2DBody()}
     *
     * @param rotation
     */
    void setRotationR(float rotation);

    Body getB2DBody();

    Transform getTransform();

    public static class B2DPhysicsRoot implements PhysicsRoot {
        Body body;

        public B2DPhysicsRoot(Body body) {
            this.body = body;
        }

        @Override
        public Vector2 position() {
            return body.getPosition();
        }

        @Override
        public float rotationR() {
            return body.getAngle();
        }

        @Override
        public void setPosition(float x, float y) {
            body.setTransform(x, y, body.getAngle());
        }

        @Override
        public void setRotationR(float rotation) {
            Vector2 position = body.getPosition();
            body.setTransform(position.x, position.y, rotation);
        }

        @Override
        public Body getB2DBody() {
            return body;
        }

        @Override
        public Transform getTransform() {
            return body.getTransform();
        }
    }

    public static class BasicPhysics implements PhysicsRoot {
        Vector2 position = N.ew(Vector2.class);
        float rotationR = 0;
        Transform transform = new Transform();

        @Override
        public Vector2 position() {
            return position;
        }

        @Override
        public float rotationR() {
            return rotationR;
        }

        @Override
        public void setPosition(float x, float y) {
            position.set(x, y);
        }

        @Override
        public void setRotationR(float rotation) {
            rotationR = rotation;
        }

        @Override
        public Body getB2DBody() {
            return null;
        }

        @Override
        public Transform getTransform() {
            transform.setPosition(position);
            transform.setRotation(rotationR);
            return transform;
        }
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/physics/PhysicsRoot.kt
package com.binarymonks.jj.core.physics

import box2dLight.Light
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.Transform
import com.binarymonks.jj.core.JJ
import com.binarymonks.jj.core.things.Thing
import com.binarymonks.jj.core.utils.NamedArray

class PhysicsRoot(val b2DBody: Body) {
    var parent: Thing? = null
        set(value) {
            field = value
            b2DBody.userData = value
            collisionResolver.parent = value
        }
    var collisionResolver: CollisionResolver = CollisionResolver()
    var lights: NamedArray<Light> = NamedArray()

    fun position(): Vector2 {
        return b2DBody.position
    }

    fun rotationR(): Float {
        return b2DBody.angle
    }

    fun setPosition(x: Float, y: Float) {
        b2DBody.setTransform(x, y, b2DBody.angle)
    }

    fun setPosition(position: Vector2) {
        b2DBody.setTransform(position.x, position.y, b2DBody.angle)
    }

    fun setRotationR(rotation: Float) {
        val position = b2DBody.position
        b2DBody.setTransform(position.x, position.y, rotation)
    }

    val transform: Transform
        get() = b2DBody.transform

    fun hasProperty(propertyKey: String): Boolean {
        if (parent != null) return parent!!.hasProp(propertyKey)
        return false
    }

    internal fun destroy(pooled: Boolean) {
        JJ.B.physicsWorld.b2dworld.destroyBody(b2DBody)
    }

}

